eng/common/pipelines/templates/steps/cosmos-emulator.yml

```diff
@@ -3,93 +3,34 @@ parameters:
-        }
-        default {
-          Write-Host "Unrecognized exit code $process.ExitCode"
-          return
-        }
-      }
-      $argumentList = ""
-      if (-not [string]::IsNullOrEmpty("${{ parameters.StartParameters }}")) {
-          $argumentList += , "${{ parameters.StartParameters }}"
-      } else {
-        # Use the default params if none provided
-        $argumentList = "/noexplorer /noui /enablepreview /disableratelimiting /enableaadauthentication"
-      }
-      Write-Host "Starting emulator process: $Emulator $argumentList"
-      $process=Start-Process $Emulator -ArgumentList $argumentList -ErrorAction Stop -PassThru
-      Write-Host "Emulator process started: $($process.Name), $($process.FileVersion)"
-      $Timeout = 600
-      $result="NotYetStarted"
-      $complete = if ($Timeout -gt 0) {
-        $start = [DateTimeOffset]::Now
-        $stop = $start.AddSeconds($Timeout)
-        {
-          $result -eq "Running" -or [DateTimeOffset]::Now -ge $stop
-        }
-      }
-      else {
-        {
-          $result -eq "Running"
-        }
-      }
-      do {
-        $process = Start-Process $Emulator -ArgumentList "/getstatus" -PassThru -Wait
-        switch ($process.ExitCode) {
-          1 {
-            Write-Host "The emulator is starting"
-          }
-          2 {
-            Write-Host "The emulator is running"
-            $result="Running"
-            return
-          }
-          3 {
-            Write-Host "The emulator is stopped"
-          }
-          default {
-            Write-Host "Unrecognized exit code $process.ExitCode"
-          }
-        }
-        Start-Sleep -Seconds 5
-      }
-      until ($complete.Invoke())
-      Write-Error "The emulator failed to reach Running status within ${Timeout} seconds"
-    displayName: Start Cosmos DB Emulator
\ No newline at end of file
+  - task: Powershell@2
+    inputs:
+      filePath: $(Build.SourcesDirectory)/eng/common/scripts/Cosmos-Emulator.ps1
+      arguments: >
+        -EmulatorMsiUrl "${{ parameters.EmulatorMsiUrl }}"
+        -StartParameters "${{ parameters.StartParameters }}"
+        -Stage "Install"
+      pwsh: true
+    displayName: Install Public Cosmos DB Emulator
+
+  - task: Powershell@2
+    inputs:
+      filePath: $(Build.SourcesDirectory)/eng/common/scripts/Cosmos-Emulator.ps1
+      arguments: >
+        -EmulatorMsiUrl "${{ parameters.EmulatorMsiUrl }}"
+        -StartParameters "${{ parameters.StartParameters }}"
+        -Stage "Launch"
+      pwsh: true
+    displayName: Launch Public Cosmos DB Emulator
+    continueOnError: true
+
+  - task: Powershell@2
+    inputs:
+      filePath: $(Build.SourcesDirectory)/eng/common/scripts/Cosmos-Emulator.ps1
+      arguments: >
+        -EmulatorMsiUrl "${{ parameters.EmulatorMsiUrl }}"
+        -StartParameters "${{ parameters.StartParameters }}"
+        -Stage "Launch"
+      pwsh: true
+    displayName: Retry Launch of Public Cosmos DB Emulator
```

Comment:
I'm not exactly sure how this "retry" logic will work...

If the emulator is already in the process of being started and it is waiting on some MDM/firewall/policies which delays the self sign certificate being registered, or the opening of the service ports, or even the creation of the local emulator data, then the best way to ensure proper retry is to kill the ongoing processes before retrying.

Alternatively calling the emulator with "/shutdown" option should take care of properly shutting down the current instance. But there might be cases where the shutdown might not be always work if the running emulator is stuck waiting on some system related call to return.

Reply:
I've reverted this in my PR https://github.com/Azure/azure-sdk-tools/pull/3509 as we shouldn't need an explicit retry step we should just handle it in the script if needed.